R/oauth-flow-device.R

```diff
@@ -49,20 +54,33 @@ req_oauth_device <- function(req, client,
 #' @keywords internal
 oauth_flow_device <- function(client,
                               auth_url,
+                              pkce = FALSE,
                               scope = NULL,
                               auth_params = list(),
                               token_params = list()) {
   oauth_flow_check("device", client, interactive = TRUE)
 
+  if (pkce) {
+    code <- oauth_flow_auth_code_pkce()
+    auth_params$code_challenge <- code$challenge
+    auth_params$code_challenge_method <- code$method
+    token_params$code_verifier <- code$verifier
+  }
+
   request <- oauth_flow_device_request(client, auth_url, scope, auth_params)
 
   # User interaction
   # https://datatracker.ietf.org/doc/html/rfc8628#section-3.3
   # Azure provides a message that we might want to print?
   # Google uses verification_url instead of verification_uri
+  # verification_uri_complete is optional, it would ship the user
+  # code in the uri https://datatracker.ietf.org/doc/html/rfc8628#section-3.2
   if (is_interactive() && has_name(request, "verification_uri_complete")) {
     inform(glue("Use code {request$user_code}"))
     utils::browseURL(request$verification_uri_complete)
+  } else if (is_interactive() && has_name(request, "verification_uri")) {
```

Comment:
The problem with informing the user about the code and then immediately calling `browseURL` is that focus will switch to the browser before they have a chance to see the code in RStudio.

Reply:
Yes, this is not ideal, indeed. However this was also the prior workflow. From a user perspective I find it easier to switch back from the browser to RStudio, where I know that I had made the call, rather than to look for the browser and potentially the tab, where I have to paste the device code into. This gets even more difficult if you have to C&P the URL also and the device code. I mean usually you will be asked in the device code flow to type in the device code. Then you start looking where to find the code and that would be where you had made the call.

Potential solutions to this:
* delaying the `browseURL` call with a timeout
* adding more informative text, how the user shall use the device in either the Rd file or directly printed to console

But this gets off-topic a bit. I mean, the intention of this PR is the enablement of PKCE in the device code flow and not so much the Human-Computer Interaction optimizations.

Reply:
Oh I see what you mean. I think the construction of the `if` statement is throwing me off. Maybe it should be more like:

```R
url <- request$verification_uri_complete %||% request$verification_uri %||% request$verification_url

if (is_interactive()) {
  inform(glue("Use code {request$user_code}"))
  utils::browseURL(request$verification_uri_complete)
} else { 
  inform(glue("Visit <{url}> and enter code {request$user_code}"))
}
```